cellxgene_schema_cli/cellxgene_schema/validate.py

```diff
@@ -1112,41 +1119,67 @@ def _get_raw_x_loc(self) -> str:
         else:
             return "X"
 
-    def _is_raw(self, force: bool = False) -> bool:
+    def _has_valid_raw(self, force: bool = False) -> bool:
         """
         Checks if the non-zero values for the raw matrix (adata.X or adata.raw.X)
-        are integers. Returns False if at least one value is not an integer,
+        are positive integers stored as numpy.float32. Also validates that every row contains at least one non-zero
+        value.
 
-        True otherwise.
+        Returns False if at least one value / row does not meet requirements. True otherwise.
 
         Since this process is memory intensive, it will return a cache value if this function has been called before.
         If calculation needs to be repeated use `force = True`
 
         :rtype bool
-        :return False if at least one value is not an integer, True otherwise
         """
         if force:
             self._raw_layer_exists = None
 
         if self._raw_layer_exists is None:
             # Get potential raw_X
-            raw_loc = self._get_raw_x_loc()
-            x = self.adata.raw.X if raw_loc == "raw.X" else self.adata.X
+            x = self.adata.raw.X if self._get_raw_x_loc() == "raw.X" else self.adata.X
+            if x.dtype != np.float32:
+                self._raw_layer_exists = False
+                self.errors.append("Raw matrix values must have type numpy.float32.")
+                return self._raw_layer_exists
 
             matrix_format = get_matrix_format(self.adata, x)
             assert matrix_format != "unknown"
             self._raw_layer_exists = True
+            has_row_of_zeros = False
+            has_invalid_nonzero_value = False
+            is_sparse_matrix = matrix_format in SPARSE_MATRIX_TYPES
             for matrix_chunk, _, _ in self._chunk_matrix(x):
-                data = matrix_chunk if isinstance(matrix_chunk, np.ndarray) else matrix_chunk.data
-                if (data % 1 > 0).any():
-                    self._raw_layer_exists = False
+                if not has_row_of_zeros:
+                    if is_sparse_matrix:
+                        row_indices, _ = matrix_chunk.nonzero()
+                        if len(set(row_indices)) != self.adata.n_obs:
+                            has_row_of_zeros = True
+                    # else, must be dense matrix, confirm that all rows have at least 1 nonzero value
+                    elif not all(np.apply_along_axis(np.any, axis=1, arr=matrix_chunk)):
+                        has_row_of_zeros = True
+
+                if not has_invalid_nonzero_value:
+                    data = matrix_chunk if isinstance(matrix_chunk, np.ndarray) else matrix_chunk.data
+                    if np.any((data % 1 > 0) | (data < 0)):
+                        has_invalid_nonzero_value = True
+
+                if has_row_of_zeros and has_invalid_nonzero_value:
```

Comment:
```suggestion
                if has_row_of_zeros or has_invalid_nonzero_value:
```

Reply:
I decided to ask Jason about this in the [issue](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/gh/chanzuckerberg/single-cell-curation/614), since I'm not sure if we want to keep validating for one rule if the other is already found to cause an error. If he says to fail immediately on violation of either, we can even get rid of the flags altogether and just `break` once we encounter either error

Reply:
https://github.com/chanzuckerberg/single-cell-curation/issues/614#issuecomment-1743863137

Looks like we want to keep the `and`, so we validate the other rule even if one has already failed, and we break if both rules have errored